### PR TITLE
ISSUE-1088 TeamMailboxMember/get JMAP method

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+      .addUser(ANDRE.asString(), ANDRE_PASSWORD)
+
+    requestSpecification = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .build()
+  }
+
+  @Test
+  def missingPublicAssetCapabilityShouldFail(): Unit =
```
Comment (Arsnael) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala:35]: bad copy paste, should be `missingTeamMailboxesCapabilityShouldFail`

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      .collectSeq()
+      .flatMap(teamMailboxes => {
+        val foundTeamMailboxes = teamMailboxes.filter(teamMailbox => mailboxNames.contains(teamMailbox.mailboxName.asString()))
+        val notFoundTeamMailboxNames = mailboxNames.diff(foundTeamMailboxes.map(teamMailbox => teamMailbox.mailboxName.asString()).toSet).toSeq
+        SFlux.fromIterable(foundTeamMailboxes)
+          .flatMap(teamMailbox =>
+            SFlux(teamMailboxRepository.listMembers(teamMailbox))
+              .collectMap(member => member.username.asString(), member => TeamMailboxMemberRoleDTO(member.role.value.toString))
+              .map(mapMembers => TeamMailboxMemberDTO(teamMailbox.mailboxName.asString(), mapMembers)))
+          .collectSeq()
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala:73]: From the `flatMap(teamMailbox`  to the collectSeq, duplicated with def above. Extract it into a private def

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      case Some(ids) => getMembersOfSpecificTeamMailboxes(username, request.accountId, ids)
+    }
+  }
+
+  private def getMembersOfAllTeamMailboxes(username: Username, accountId: AccountId): SMono[TeamMailboxMemberGetResponse] =
+    SFlux.fromPublisher(teamMailboxRepository.listTeamMailboxes(username))
+      .flatMap(teamMailbox =>
+        SFlux(teamMailboxRepository.listMembers(teamMailbox))
+          .collectMap(member => member.username.asString(), member => TeamMailboxMemberRoleDTO(member.role.value.toString))
+          .map(mapMembers => TeamMailboxMemberDTO(teamMailbox.mailboxName.asString(), mapMembers)))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala:58]: Extract this in a sub method and name it please

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    SFlux.fromPublisher(teamMailboxRepository.listTeamMailboxes(username))
+      .collectSeq()
+      .flatMap(teamMailboxes => {
+        val foundTeamMailboxes = teamMailboxes.filter(teamMailbox => mailboxNames.contains(teamMailbox.mailboxName.asString()))
+        val notFoundTeamMailboxNames = mailboxNames.diff(foundTeamMailboxes.map(teamMailbox => teamMailbox.mailboxName.asString()).toSet).toSeq
+        SFlux.fromIterable(foundTeamMailboxes)
+          .flatMap(teamMailbox =>
+            SFlux(teamMailboxRepository.listMembers(teamMailbox))
+              .collectMap(member => member.username.asString(), member => TeamMailboxMemberRoleDTO(member.role.value.toString))
+              .map(mapMembers => TeamMailboxMemberDTO(teamMailbox.mailboxName.asString(), mapMembers)))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala:72]: Duplicated code with (a)

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+import net.javacrumbs.jsonunit.core.Option.IGNORING_ARRAY_ORDER
+import org.apache.http.HttpStatus.SC_OK
+import org.apache.james.GuiceJamesServer
+import org.apache.james.jmap.core.ResponseObject.SESSION_STATE
+import org.apache.james.jmap.http.UserCredential
+import org.apache.james.jmap.rfc8621.contract.Fixture.{ACCEPT_RFC8621_VERSION_HEADER, ANDRE, ANDRE_PASSWORD, BOB, BOB_PASSWORD, DOMAIN, authScheme, baseRequestSpecBuilder}
+import org.apache.james.utils.DataProbeImpl
+import org.junit.jupiter.api.{BeforeEach, Test}
+
+trait TeamMailboxMemberGetMethodContract {
```
Comment (chibenwa) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala:19]: Please add a test: bob tries to list members of a team mailbox he do not have access to.

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/TeamMailboxMemberGetRequest.scala [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.linagora.tmail.james.jmap.model
+
+import org.apache.james.jmap.core.AccountId
+import org.apache.james.jmap.method.WithAccountId
+
+case class TeamMailboxMemberGetRequest(accountId: AccountId,
+                                 ids: Option[Set[String]] = None) extends WithAccountId
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/model/TeamMailboxMemberGetRequest.scala:7]: indent

**Review by Arsnael — APPROVED** [review]

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+           |    }, "c1"]
+           |  ]
+           |}""".stripMargin))
+
+  @Test
+  def getShouldReturnEmptyListByDefault(): Unit =
+    `given`
+      .body(
+        s"""{
+           |	"using": ["urn:ietf:params:jmap:core", "com:linagora:params:jmap:team:mailboxes"],
```
Comment (quantranhong1999) [inline tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala:109]: missing the `urn:ietf:params:jmap:mail` capability

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        arguments = Arguments(Serializer.serializeGetResponse(response).as[JsObject]),
+        methodCallId = invocation.invocation.methodCallId))
+      .map(InvocationWithContext(_, invocation.processingContext))
+
+  private def getTeamMailboxMemberResponse(username: Username, request: TeamMailboxMemberGetRequest): SMono[TeamMailboxMemberGetResponse] = {
+    request.ids match {
+      case None => getMembersOfAllTeamMailboxes(username, request.accountId)
+      case Some(ids) => getMembersOfSpecificTeamMailboxes(username, request.accountId, ids)
+    }
+  }
```
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala:49]: This bracket can be removed.

File: tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala (hunk 7 above)
Comment (Arsnael) [reply tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/TeamMailboxMemberGetMethodContract.scala:109]: Ah you are right actually, we use it for revoke access method as well

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala [hunk 9]
```diff
@@ -41,31 +41,19 @@ class TeamMailboxMemberGetMethod @Inject()(val teamMailboxRepository: TeamMailbo
         methodCallId = invocation.invocation.methodCallId))
       .map(InvocationWithContext(_, invocation.processingContext))
 
-  private def getTeamMailboxMemberResponse(username: Username, request: TeamMailboxMemberGetRequest): SMono[TeamMailboxMemberGetResponse] = {
+  private def getTeamMailboxMemberResponse(username: Username, request: TeamMailboxMemberGetRequest): SMono[TeamMailboxMemberGetResponse] =
     request.ids match {
-      case None => getMembersOfAllTeamMailboxes(username, request.accountId)
-      case Some(ids) => getMembersOfSpecificTeamMailboxes(username, request.accountId, ids)
```
Comment (Arsnael) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala]: Why changing this? If anything I think it's harder to read now

Comment (hungphan227) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberGetMethod.scala]: It is suggestion from Tung. Fixed

**Review by Arsnael — APPROVED** [review]

**Pull request merged**